# <center>Profilerstellung von Wählern der demokratischen und der republikanischen Parteien</center>

<div style="text-align: justify"> Da ich mich zurzeit mit dem Thema personalisierte Wähleransprache (political Microtargeting) beschäftige, hatte ich die Idee parallel zu der Verfassung meiner Magisterarbeit auch eine Analyse der Wähler der größten amerikanischen Parteien zu machen. Mein persönlicher Beweggrund dafür war etwas konkreter zu zeigen, wie man Erkenntnisse aus Wählerdaten ziehen kann. Somit will ich auch eine gewisse "Lücke" füllen, denn im Internet findet man unterschiedliche Informationen aus unterschiedlichen Quellen über was Politiker mit Microtargeting erreichen können, aber wie die Daten ausgewertet werden, um dieses Ziel zu erreichen, ist oft unersichtlich.</div>  
    
Aus diesem Grund will ich mit diesem Projekt diesen "Zwischenschritt" zeigen, damit wir eine bessere Vorstellung bekommen können, wie die Auswertung der personenbezogenen Wählerdaten erfolgt.

# <center> Ziel </center>

Das Ziel der Analyse ist folgende Fragen zu beantworten:

1) Was ist das Profil der Wähler der "Democratic Party" und der "Republican Party" ?

- 1.1) Was sind die Themen, die diesen Gruppen am Herzen liegen?

2) Wer sind die unentschlossene Wähler?

- 2.1) Mit welchen Inhalt kann man sie werben?
    
    
3) In welchen Staaten wohnen die Anhänger der Republican Party? und der Democratic Party?

- 3.1) In welchen Staaten wohnen die unentschlossene Wähler?

Diese Fragen nehmen Bezug zu Artikeln, die ich gelesen habe im Rahmen meiner Magisterarbeit. [Manche Autoren zeigen, dass die erfolgreiche Ansprache der überzeugten Wähler ist der Schlüsselfaktor für den Wahlsieg](https://journals.sagepub.com/doi/abs/10.1177/1354068815605676)(Frage 1). Andere Aufsätze sind eher der Meinung, dass das wahre Potenzial des Microtargetings liegt an [der Möglichkeit, unentschlossene Wähler anzusprechen](https://www.utrechtlawreview.org/articles/abstract/10.18352/ulr.420/) (Frage 2). Die Dritte Frage nimmt nun Bezug zu der Kostenverwaltung während der Wahlkampagne. In den Vereinigten Staaten gibt es die sogenannte "Swing States", in denen es sehr schwierig ist, eine Mehrheit für den einen oder den anderen Politiker zu sehen. Wegen dieses Merkmals können diese Staaten besonders wichtig für das Wahlkampfergebnis sein.

Wenn diese drei Fragen beantwortet werden, können politische Partei Vorteile aus den Erkenntnissen der Daten ziehen.

# <center> Die Daten </center>

 Um die oben genannten Fragen zu beantworten, werde ich die Daten aus [Voter Study Group](https://www.voterstudygroup.org/data) analysieren. Die Gruppe hat 2016 nach der Präsidentschaftswahl eine Umfrage durchgeführt, um die Wahlpräferenzen der Amerikaner herauszufinden. Der Datensatz beinhaltet die Angaben von 8.000 Wähler aus den Vereinigten Staaten in Bezug auf ihre Parteiaffinität, Wahlverhaten, Werten usw.

Die Daten, welche normalerweise gesammelt werden, um z.B. Microtargeting über soziale Netzwerke zu betreiben sind nicht so ersichtlich. Sie stammen aus mehreren Quellen und müssen intensiv gepflegt und verarbeitet werden, bevor sie analysiert werden können. Die Daten in dieser Weise zu sammeln würde dennoch meine finanziellen und zeitlichen Kapazitäten bei weitem überstreiten. Deswegen habe ich mich entschieden, den Datensatz von Voter Study Group zu benutzen, welcher der strukturierten Form der Wählerdaten (nach Erhebung, Aufbereitung, Bereinigung usw) ähneln soll.

Darüber hinaus falle ich kein Urteil über die statistischen Methoden, die benutzt wurden, um die Daten zu erheben. In dieser Hinsicht gehe ich hier nicht auf Themen ein, wie die Representativität der Daten. Es heißt, ob der Datensatz (Probe) in der Lage ist, die Merkmale der gesamten Population zu wiederspigeln.

Mehr Information zur [Methodik der Study Group](https://www.voterstudygroup.org/publication/2016-voter-survey) kann man dennoch hier sehen.

## <center> Überblick des Datensatzes </center>

In [1]:
# Importiert die erforderlichen Libraries
import pandas as pd
import altair as alt
import geopandas as gpd
import folium
import json

survey = pd.read_csv(r"C:\Users\celio\Data Analysis\Projects\Random\Voter_Survey\VOTER_Survey_December16_Release1.csv",
                    low_memory = False)
survey.info() # zeigt grundlegende Informationen zum Datensatz

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Columns: 668 entries, case_identifier to endtime_2016
dtypes: float64(30), int64(5), object(633)
memory usage: 40.8+ MB


Wir können nun sehen, dass es sich um einen Datensatz handelt, der die Angaben von 8000 Wähler gesammelt hat. Der DataFrame hat zudem 668 Spalten. Nicht alle Spalten sind tatsächliche Fragen, aber wir können davon ausgehen, dass der Datensatz sehr umfangreich ist. 

Unten können wir sehen wie die Daten aussehen.

In [2]:
survey.head()

case_identifier    weight PARTY_AGENDAS_rand_2016  \
0              779  0.358213        Republican Party   
1             2108  0.562867        Republican Party   
2             2597  0.552138        Republican Party   
3             4148  0.207591        Democratic Party   
4             4460  0.333729        Republican Party   

           pp_primary16_2016 pp_demprim16_2016 pp_repprim16_2016  \
0  In the Democratic primary   Hillary Clinton               NaN   
1  In the Republican primary               NaN      Donald Trump   
2  In the Democratic primary   Hillary Clinton               NaN   
3  In the Democratic primary      Someone else               NaN   
4  In the Republican primary               NaN       Marco Rubio   

  inputstate_2016  izip_2016 votereg2_2016 votereg_f_2016  ...  \
0      California      94952           Yes            Yes  ...   
1         Arizona      85298           Yes            Yes  ...   
2       Wisconsin      54904           Yes            Yes  ...   
3        Oklahoma      74104           Yes            Yes  ...   
4           Texas      78253           Yes            Yes  ...   

   post_HouseCand3Name_2012 post_HouseCand3Party_2012 post_SenCand1Name_2012  \
0                       NaN                       NaN        Shelley Berkley   
1                       NaN                       NaN        Richard Carmona   
2                       NaN                       NaN          Tammy Baldwin   
3                       NaN                       NaN                    NaN   
4                       NaN                       NaN            Paul Sadler   

  post_SenCand1Party_2012 post_SenCand2Name_2012 post_SenCand2Party_2012  \
0              Democratic            Dean Heller              Republican   
1              Democratic             Jeff Flake              Republican   
2              Democratic         Tommy Thompson              Republican   
3                     NaN                    NaN                     NaN   
4              Democratic               Ted Cruz              Republican   

  post_SenCand3Name_2012 post_SenCand3Party_2012      starttime_2016  \
0                    NaN                     NaN  29nov2016 22:59:43   
1                    NaN                     NaN  29nov2016 15:41:28   
2                    NaN                     NaN  29nov2016 16:08:39   
3                    NaN                     NaN  14dec2016 18:46:33   
4                    NaN                     NaN  01dec2016 10:17:47   

         endtime_2016  
0  29nov2016 23:28:24  
1  29nov2016 18:58:28  
2  29nov2016 16:32:43  
3  14dec2016 19:11:20  
4  01dec2016 10:59:48  

[5 rows x 668 columns]

# <center> Allgemeine Analyse </center>

Um uns an den Datensatz zu gewöhnen, können wir zuerst eine allgemeine Auswertung der Daten durchführen. Das Ziel dieses Teils der Analyse ist folgende Fragen zu beanworten:

1) Wie viele Wähler jeder Partei gibt es im Datensatz?

2) Wie sind die Wähler durch das Land verteilt?

3) Sind Republikaner und Demokraten unterschiedlich engagiert im Wahlkampf?

4) Was sind die Anliegen der Wäher der Democratic Party und der Republican Party?

## 1. Anzahl der Wäler jeder Partei

Die "PARTY_AGENDAS_rand_2016" Spalte zeigt welche Umfrageteilnehmer sich der einen oder der anderen Partei affin erklärt haben. Wenn wir eine Häufigkeitstabelle erstellen, können wir sehen, wie viele Wähler von jeder Partei es im Datensatz gibt.

In [421]:
survey["PARTY_AGENDAS_rand_2016"].value_counts()

Republican Party    4108
Democratic Party    3892
Name: PARTY_AGENDAS_rand_2016, dtype: int64

## 2. Verteilung der Wähler nach Staat

Es gibt zwei Spalten, die uns helfen, diese Information zu finden, nämlich "inputstate_2016" und "PARTY_AGENDAS_rand_2016". Wenn wir die Teilnehmer nach den Variablen in diesen zwei Spalten gruppieren und die Größe der Gruppe errechnen, finden wir wie viele Wähler gibt es insgesament und in jedem Staat.

In [3]:
verteilung = survey.copy() # kopiert den ursprünglichen Datensatz, um die Manipulation von den originalen Daten zu vermeiden
verteilung = verteilung.pivot_table(index=["inputstate_2016","PARTY_AGENDAS_rand_2016"],aggfunc="size").unstack()
# Erstellt ein Pivot mit den Spalten "inputstate_2016" und "PARTY_AGENDAS_rand_2016". "size" errechnet die Größe

total = pd.DataFrame(index = ["Total"],data = {"Democratic Party":verteilung["Democratic Party"].sum(),
                                              "Republican Party":verteilung["Republican Party"].sum()})
# Instanziiert ein DataFrame mit den Gesamtsummen der Wähler jeder Partei
verteilung_margins = pd.concat([verteilung,total]) # fügt die neue DataFrame zur Pivot Table hinzu.
verteilung_margins["dem_percentage(%)"]=verteilung_margins["Democratic Party"]/verteilung_margins["Democratic Party"].sum()*100
verteilung_margins["rep_percentage(%)"]=verteilung_margins["Republican Party"]/verteilung_margins["Republican Party"].sum()*100
# Die beiden Zeilen oben fügen zwei Spalten hinzu, wo man sehen kann, wie viel jeder Staat beiträgt
# zu der gesamten Anzahl von Wählern
verteilung_margins

PARTY_AGENDAS_rand_2016  Democratic Party  Republican Party  \
Alabama                                40                40   
Alaska                                  8                13   
Arizona                               105               118   
Arkansas                               29                51   
California                            375               435   
Colorado                               81                68   
Connecticut                            35                42   
Delaware                               19                11   
District of Columbia                   10                10   
Florida                               324               328   
Georgia                               126               116   
Hawaii                                 16                 9   
Idaho                                  33                33   
Illinois                              134               182   
Indiana                                78                89   
Iowa                                   40                53   
Kansas                                 40                45   
Kentucky                               41                61   
Louisiana                              32                38   
Maine                                  28                22   
Maryland                               72                80   
Massachusetts                          74                99   
Michigan                              126               147   
Minnesota                              79                52   
Mississippi                            25                25   
Missouri                               78                88   
Montana                                18                10   
Nebraska                               29                23   
Nevada                                 44                38   
New Hampshire                          21                30   
New Jersey                            100                99   
New Mexico                             30                30   
New York                              198               217   
North Carolina                        114               112   
North Dakota                            4                 9   
Ohio                                  153               153   
Oklahoma                               36                42   
Oregon                                 78                64   
Pennsylvania                          226               215   
Rhode Island                           19                14   
South Carolina                         47                51   
South Dakota                           20                 9   
Tennessee                              70                75   
Texas                                 242               270   
Utah                                   44                31   
Vermont                                13                 9   
Virginia                              110               100   
Washington                            109               125   
West Virginia                          26                33   
Wisconsin                              86                83   
Wyoming                                 7                11   
Total                                3892              4108   

PARTY_AGENDAS_rand_2016  dem_percentage(%)  rep_percentage(%)  
Alabama                           0.513875           0.486855  
Alaska                            0.102775           0.158228  
Arizona                           1.348921           1.436222  
Arkansas                          0.372559           0.620740  
California                        4.817575           5.294547  
Colorado                          1.040596           0.827653  
Connecticut                       0.449640           0.511198  
Delaware                          0.244090           0.133885  
District of Columbia              0.128469           0.121714  
Florida             

Um diese Informationen visuell interessanter zu machen, können wir sie mithilfe einer Grafik darstellen. In der Grafik unten können Sie auf den Namen der Staaten klicken, um die Anzahl der Wähler in diesem Staat zu sehen. Wenn Sie beim klicken auf die Taste "shift" halten, können Sie mehrere Orte wählen.

In [4]:
verteilung_selection = alt.selection_multi(fields = ["index"],nearest = True)

verteilung_chart_left=alt.Chart(verteilung_margins.iloc[:-1,:].reset_index()).mark_bar().encode(
    x= alt.X("Democratic Party",sort = "descending", title = "Number of Voters"),
    y= alt.Y("index", axis = alt.Axis(labels = False,title = "S t a t e s", titleFontSize=12)),
    color = alt.value("blue")).transform_filter(verteilung_selection).properties(width = 250)

verteilung_chart_left_text = verteilung_chart_left.mark_text(align ="right",dx = -10).encode(
    text = "index")

verteilung_dem = (verteilung_chart_left+verteilung_chart_left_text).add_selection(verteilung_selection)

verteilung_chart_right=alt.Chart(verteilung_margins.iloc[:-1,:].reset_index()).mark_bar().encode(
    x = alt.X("Republican Party", title = "Number of Voters"),
    y = alt.Y("index", axis = alt.Axis(labels=False, title = "S t a t e s", titleFontSize=12)),
    color = alt.value("firebrick")).transform_filter(verteilung_selection).properties(width=250)

verteilung_chart_right_text = verteilung_chart_right.mark_text(align = "left",dx = 10).encode(
    text = "index")

verteilung_rep = (verteilung_chart_right+verteilung_chart_right_text).add_selection(verteilung_selection)

verteilung_chart_middle = alt.Chart(verteilung_margins.iloc[:-1,:].reset_index()).mark_point(align = "right").encode(
    y= alt.Y("index",title = ""),
    color = "index").add_selection(verteilung_selection)

alt.hconcat(verteilung_dem,verteilung_chart_middle,verteilung_rep)

alt.HConcatChart(...)

## 2. Engagement im Wahlkampf (Wahlbeteiligung)

Engagieren sich Republikaner und Demokraten ähnlich in der Wahl?

Unten rechnen wir mithilfe der Spalte "turnout16_2016" wie viele Wähler von jeder Partei tatsächlich an der Präsidentschaftswahl 2016 teilgenommen haben.

In [5]:
wahlbeteiligung = survey.pivot_table(index="PARTY_AGENDAS_rand_2016",columns = "turnout16_2016",aggfunc="size")
wahlbeteiligung["percentage_nicht_wähler"] = [(154/(154+3713)),(187/(187+3898))]
wahlbeteiligung

turnout16_2016            No   Yes  percentage_nicht_wähler
PARTY_AGENDAS_rand_2016                                    
Democratic Party         154  3713                 0.039824
Republican Party         187  3898                 0.045777

In [422]:
wahlbeteiligung_chart_data=wahlbeteiligung.reset_index().melt(id_vars=["PARTY_AGENDAS_rand_2016","percentage_nicht_wähler"],
                                                                         value_vars=["No","Yes"])
wahlbeteiligung_dem=wahlbeteiligung_chart_data.loc[wahlbeteiligung_chart_data["PARTY_AGENDAS_rand_2016"]=="Democratic Party"]
wahlbeteiligung_dem_chart = alt.Chart(wahlbeteiligung_dem,
                                     title = "Democratic Party").mark_bar().encode(
    x = alt.X("turnout16_2016", axis = alt.Axis(labelAngle =0,
                                               title = "Haben Sie an der 2016 Präsidentschaftswahl teilgenommen?")),
    y = "value",
    color = "turnout16_2016").properties(width=300)

wahlbeteiligung_rep=wahlbeteiligung_chart_data.loc[wahlbeteiligung_chart_data["PARTY_AGENDAS_rand_2016"]=="Republican Party"]
wahlbeteiligung_rep_chart = alt.Chart(wahlbeteiligung_rep,
                                     title = "Republican Party").mark_bar().encode(
    x = alt.X("turnout16_2016", axis = alt.Axis(labelAngle =0,
                                                title = "Haben Sie an der 2016 Präsidentschaftswahl teilgenommen?")),
    y = "value",
    color = "turnout16_2016").properties(width=300)

wahlbeteiligung_dem_chart|wahlbeteiligung_rep_chart

alt.HConcatChart(...)

Wir können sehen, dass die Wahlbeteiligung unter den Demokraten ein bisschen höher war als bei den Republikanern. Das kann dennoch daran liegen, dass es mehr Republikaner an der Umfrage teilgenommen haben (je mehr teilnehmer, desto größer die Anzahl der nicht-wähler). Zudem ist das Verhältnis zwischen Wählern und nicht Wählern ungefähr das Gleich bei beiden Parteien, ohne dass wir einen Unterschied sehen können in Bezug auf das Engagement der Wähler der einen oder der anderen Partei. 

## 4. Wählerprofil

Die Spalten mit den Namen "PARTY_AGENDAS_D1_2016" bis "PARTY_AGENDAS_D12_2016" enthalten Informationen über die Erwartungen der Wähler gegenüber der Partei, die sie gewählt haben. Die Spalten "PARTY_AGENDAS_R1_2016" bis "PARTY_AGENDAS_R12_2016" enthalten dieselben Informationen, aber bezüglich der republikanischen Partei.

Wir können die Spalten umbenennen, damit wir wissen können, worum es in jeder Spalte geht. Eine Liste mit den Namen der Spalten kann auf der [Webseite von Voter Survey](https://www.voterstudygroup.org/publication/2016-voter-survey) gefunden werden.

Zudem können wir die Umfrageteilnehmer nach Antwortart gruppieren, so dass wir anfangen können ein Profil zu skizzieren.

In [520]:
d = survey.loc[survey["PARTY_AGENDAS_rand_2016"]=="Democratic Party"] # Schneidet die DF für "Democratic Party"
r = survey.loc[survey["PARTY_AGENDAS_rand_2016"]=="Republican Party"] # Scheneidet die DF für "Republican Party"
d_col_names = ["PARTY_AGENDAS_rand_2016"] # Instanziiert eine List für die Namen der Spalten
r_col_names = ["PARTY_AGENDAS_rand_2016"] # Instanziiert eine List für die Namen der Spalten
for n in range(1,13):
    col_name_d = "PARTY_AGENDAS_D{}_2016".format(n)
    col_name_r = "PARTY_AGENDAS_R{}_2016".format(n)
    d_col_names.append(col_name_d)
    r_col_names.append(col_name_r)
    
d = d[d_col_names]
new_col_names = ["PARTY_AGENDAS_rand_2016","reducing_size_government","creating_jobs","renegotiating_trade_deals",
             "reforming_health_care","reducing_taxes","combating_climate_change","reducing_poverty",
             "restricting_immigration","reducing_national_debt","increasing_racial_equality",
             "Combating_terrorism","political_correctness"]
d.columns = new_col_names
r = r[r_col_names]
r.columns = new_col_names
d_melted = d.melt(id_vars ="PARTY_AGENDAS_rand_2016",value_name="Importance",var_name="Matter")
d_melted = d_melted.loc[d_melted["Importance"].notnull()]
r_melted = r.melt(id_vars="PARTY_AGENDAS_rand_2016",value_name="Importance",var_name="Matter")
r_melted = r_melted.loc[r_melted["Importance"].notnull()]
both_melted = pd.concat([d_melted,r_melted])
both_pivot = both_melted.pivot_table(index=["PARTY_AGENDAS_rand_2016","Matter","Importance"],
                        aggfunc="size")
both_pivot_table = both_pivot.unstack()
both_pivot_table = both_pivot_table[["Much more","Somewhat more",
                                     "About the same","Somewhat less",
                                     "Much less"]]

for party in ["Democratic Party","Republican Party"]:
    group = both_pivot_table.loc[party]
    for matter in group.index:
        total = group.loc[matter].sum()
        both_pivot_table.loc[party,matter] = (both_pivot_table.loc[party,matter]/total*100)
        
both_pivot_table = both_pivot_table.applymap("{:.2f}%".format)


Importance                                         Much more Somewhat more  \
PARTY_AGENDAS_rand_2016 Matter                                               
Democratic Party        Combating_terrorism           39.07%        21.86%   
                        combating_climate_change      23.66%        18.13%   
                        creating_jobs                 50.47%        30.30%   
                        increasing_racial_equality    19.80%        20.64%   
                        political_correctness         30.89%        17.42%   
                        reducing_national_debt        45.74%        23.76%   
                        reducing_poverty              32.27%        28.31%   
                        reducing_size_government      40.07%        19.12%   
                        reducing_taxes                39.51%        24.69%   
                        reforming_health_care         42.60%        26.04%   
                        renegotiating_trade_deals     26.54%        28.80%   
                        restricting_immigration       37.50%        16.28%   
Republican Party        Combating_terrorism           37.65%        21.74%   
                        combating_climate_change      32.33%        14.52%   
                        creating_jobs                 52.06%        29.22%   
                        increasing_racial_equality    30.17%        19.63%   
                        political_correctness         31.72%        18.21%   
                        reducing_national_debt        40.64%        27.80%   
                        reducing_poverty              40.23%        29.22%   
                        reducing_size_government      34.46%        20.42%   
                        reducing_taxes                33.70%        26.13%   
                        reforming_health_care         40.64%        21.89%   
                        renegotiating_trade_deals     23.28%        28.56%   
                        restricting_immigration       33.30%        17.77%   

Importance                                         About the same  \
PARTY_AGENDAS_rand_2016 Matter                                      
Democratic Party        Combating_terrorism                31.69%   
                        combating_climate_change           23.05%   
                        creating_jobs                      15.33%   
                        increasing_racial_equality         30.49%   
                        political_correctness              25.58%   
                        reducing_national_debt             22.90%   
                        reducing_poverty                   25.51%   
                        reducing_size_government           27.73%   
                        reducing_taxes                     24.48%   
                        reforming_health_care              16.06%   
                        renegotiating_trade_deals          32.84%   
                        restricting_immigration            27.59%   
Republican Party        Combating_terrorism                30.89%   
                        combating_climate_change           20.24%   
                        creating_jobs                      15.23%   
                        increasing_racial_equality         31.93%   
                        political_correctness              19.50%   
                        reducing_national_debt             19.99%   
                        reducing_poverty                   22.25%   
                        reducing_size_government           22.20%   
                        reducing_taxes                     22.03%   
                        reforming_health_care              15.15%   
                        renegotiating_trade_deals          27.22%   
                        restricting_immigration            19.31%   

Importance                                         Somewhat less Much less  
PARTY_AGENDAS_rand_2016 Matter                                              
Democratic Party      

In der Dashboard unten kann man eine bestimmte Angelegenheit wählen, indem man die Maus auf die Blocks unten setzt. Je nach Wahl ändern sich die zwei Grafiken oben, so dass sie die Präferenzen der Wähler jeder Partei widerspiegeln.

In [521]:
both_chart_data = both_pivot.reset_index().rename({0:"n_voters"},axis = "columns")
both_chart_selection = alt.binding_select(options =[x for x in both_chart_data["Matter"].unique()])
both_chart_condition = alt.selection_single(fields = ["Matter"], on = "mouseover", nearest = True)
dem_chart = alt.Chart(both_chart_data.loc[both_chart_data["PARTY_AGENDAS_rand_2016"]=="Democratic Party"],
                     title = "Democratic Party").mark_bar().encode(
    x = alt.X("Importance", axis = alt.Axis(labelAngle=-45,labelFontSize=12),
             sort = ["Much more","Somewhat more","About the same","Somewhat less","Much less","Didn't Answer"]),
    y = alt.Y("n_voters", title = "Number of Voters"),
    color = alt.Color("n_voters", 
                      scale = alt.Scale(scheme="blues"))).properties(width=300).transform_filter(both_chart_condition)

rep_chart=alt.Chart(both_chart_data.loc[both_chart_data["PARTY_AGENDAS_rand_2016"]=="Republican Party"],
                   title = "Republican Party").mark_bar().encode(
    x = alt.X("Importance",axis = alt.Axis(labelAngle=-45,labelFontSize=12),
             sort = ["Much more","Somewhat more","About the same","Somewhat less","Much less","Didn't Answer"]),
    y = alt.Y("n_voters", title = "Number of Voters"),
    color = alt.Color("n_voters",
                      scale=alt.Scale(scheme="reds"))).properties(width=300).transform_filter(both_chart_condition)

l = alt.Chart(both_chart_data).mark_bar().encode(
    x =alt.X("Matter", axis = alt.Axis(labelAngle=-45, labelFontSize=13)),
    color = alt.Color("Matter", 
                      legend= None)).properties(width=800).add_selection(both_chart_condition)

altogether = alt.hconcat(dem_chart,rep_chart).resolve_scale(color ="independent",y="shared").add_selection(both_chart_condition)
altogether&l

alt.VConcatChart(...)

Mithilfe des Dashboards können wir folgendes schließen:

**-Combating Terrorism**: sowohl Wähler der demokratischen Partei als auch Wähler der demokratischen Partei wollen, dass die Parteien sich dem Thema intensiver widmen.

**-Climate Change**: die Wähler der Republican Party wollen, dass die Partei sich mehr für Klimaschutz einsetzt. Die Demokraten sind hingegen zufriedener in Bezug auf die Einstellung der demokratischen Partei dem Thema gegenüber.

**-Creating Jobs**: beide Wählergruppen wollen, dass die Parteien sich intensiver mit dem Thema beschäftigen.

**-Increasing Racial Equality**: Wähler der Republikaner erwarten, dass die Partei sich mehr für die Gleichberechtigung einsetzt. Unter den Demokraten ist die Erwartung auch hoch, dennoch die Anzahl der Personen die "zufrieden sind" (Somewhat more und About the same" ist in dieser Gruppe auch höher).

**-Political Correctness**: Die Wähler beider Gruppen sind der Meinung, dass die Parteien sich in dieser Thematik mehr einbringen sollten.

**Reducing National Debt**: an dieser Stelle sind die Erwartungen der Demokraten größer, denn es gibt mehr Menschen die "Much more" geantwortet haben. Die Anzahl der Menschen, die dieser Meinung haben, unter den Wählern der Republican Party ist noch so hoch, andererseits gleich sich die Erwartungen aus dadurch, dass es mehr Wähler der Republikaner gibt, die "Somewhat more" geantwortet haben, als Wähler der demokratischen Partei.

**Reducing Poverty**: das Thema scheint die Republikaer stärker zu prägen als die Demokraten

**Reducing the Size of Government**: sowohl die Wähler der Democratic Party als auch die Wähler der Republikan Party wollen, dass der Statt "kleiner" wird.

**Reducing Taxes**: die zwei Gruppen zeigen ähnliche Tendenzen in Bezug auf die Senkung von Steuern. Beide Gruppen sind dafür.

**Reforming Health Care**: Auch here unterscheiden sich die Gruppen voneinander nicht. Die Wähler beider Parteien wollen, dass die Politiker sich mehr dafür einsetzen.

**Renegotiating Trade Deals**: hier sind beide Parteien eher zufrieden mit der Einstellung der Parteien in Bezug auf das Thema. Unter den republikanern gibt es denoch ein bisschen mehr Wähler, die die Meinung vertreten, die republikanische Partei sollte sich mehr dafür einsetzen.

**Restricting Immigration**: die Anzahl der Menschen, die "Much more" gewählt haben, ist in beiden Gruppen ähnlich. Die demokraten zeichnen sich dadurch aus, dass es mehr Wähler gibt, die zufrienden mit der Einstellung der Partei sind.

## Weitere Punkte

- Was ist das Profil der entschlossenen Wähler?

- Was ist das Profil der unentschlossene Wähler?

- Wo sind die überzeugten Wähler? Wo sind die überzeugbare Wähler?

# Alternative Dashboards

Da die Analysis noch nicht fertig ist, habe ich unten einige Visualisierungen erstellt, die interessant sind, um die Daten darzustellen. Ich würde mich über Feedback freuen, ob sie besser sind als diejenigen, die ich bisher benutzt habe.

In [117]:
dem_melted = dem_profil.melt(id_vars = "PARTY_AGENDAS_rand_2016",
                value_name ="Importance",
                var_name ="Matter").pivot_table(index=["Matter","Importance"],
                                                aggfunc="size").reset_index().rename({0:"n_voters"},axis = "columns")

matter_binding = alt.binding_select(options = [x for x in dem_melted["Matter"].unique()])

matter_selection = alt.selection_single(fields =["Matter"], 
                                        bind = matter_binding,
                                        name = "Choose ",
                                        empty = "all")

bars = alt.Chart(dem_melted,
                title ="Importance of Political Matters for Voters of the Democratic Party").mark_bar().encode(
    x =alt.X("Importance", 
             sort = ["Much more","Somewhat more","About the same","Somewhat less","Much less"],
             axis = alt.Axis(labelAngle=-45, labelFontSize=12)),
    y = alt.Y("n_voters", axis = alt.Axis(title = "Number of Voters", titleFontSize=12)),
    color = "n_voters")
text = bars.mark_text(dy=-10,align="center").encode(
    text = "n_voters")

(bars+text).transform_filter(matter_selection).properties(width = 600).add_selection(matter_selection)

alt.LayerChart(...)

In [52]:
matter_selection = alt.selection_single(fields = ["Matter"], on = "mouseover", nearest = True)
base = alt.Chart(dem_melted).mark_bar().encode(
    x = alt.X("Importance", 
              sort = ["Much more","Somewhat more","About the same","Somewhat less","Much less"],
             axis = alt.Axis(labelAngle=-45)),
    y = "n_voters",
    color = alt.Color("n_voters:Q", 
                      scale = alt.Scale(scheme = "blues"))).transform_filter(matter_selection).properties(width=300)

legend = alt.Chart(dem_melted).mark_bar().encode(
     y = "Matter",
    color = alt.Color("Matter", legend = None)).add_selection(matter_selection)

alt.hconcat(base,legend)

alt.HConcatChart(...)

In [137]:
rep = survey.loc[survey["PARTY_AGENDAS_rand_2016"]=="Republican Party"]
rep_cols = ["PARTY_AGENDAS_rand_2016"]
for n in range(1,13):
    col_name = "PARTY_AGENDAS_R{}_2016".format(n)
    rep_cols.append(col_name)
    
rep = rep[rep_cols]
rep.columns = new_names
rep_melted = rep.melt(id_vars="PARTY_AGENDAS_rand_2016",value_name="Importance",var_name="Matter")
rep_pivot = rep_melted.pivot_table(index=["Matter","Importance"],aggfunc="size").unstack()
rep_pivot = (rep_pivot/rep.shape[0]*100).applymap("{:.2f}%".format)
rep_pivot

Importance                 About the same Much less Much more Somewhat less  \
Matter                                                                        
combating_climate_change           19.89%    20.96%    31.77%        11.37%   
combating_terrorism                30.40%     4.04%    37.05%         5.53%   
creating_jobs                      14.97%     1.66%    51.17%         1.78%   
increasing_racial_equality         31.35%    10.44%    29.63%         7.50%   
political_correctness              19.18%    20.33%    31.21%         9.74%   
reduce_poverty                     21.84%     3.89%    39.48%         4.26%   
reduce_size_government             21.81%    12.66%    33.86%         9.86%   
reducing_national_debt             19.62%     5.53%    39.90%         5.84%   
reducing_taxes                     21.69%     9.52%    33.18%         8.35%   
reforming_health_care              14.87%    13.02%    39.90%         8.89%   
renegotiating_trade_deals          26.75%     9.32%    22.88%        11.25%   
restricting_immigration            18.94%    17.26%    32.67%        11.81%   

Importance                 Somewhat more  
Matter                                    
combating_climate_change          14.26%  
combating_terrorism               21.40%  
creating_jobs                     28.72%  
increasing_racial_equality        19.28%  
political_correctness             17.92%  
reduce_poverty                    28.68%  
reduce_size_government            20.06%  
reducing_national_debt            27.29%  
reducing_taxes                    25.73%  
reforming_health_care             21.49%  
renegotiating_trade_deals         28.07%  
restricting_immigration           17.43%

In [159]:
rep_long = rep_melted.groupby(["Matter","Importance"]).agg("size").reset_index().rename({0:"n_voters"}, axis = "columns")

rep_binding = alt.binding_select(options = [x for x in rep_long["Matter"].unique()])
rep_selection = alt.selection_single(fields =["Matter"], bind = rep_binding)

b = alt.Chart(rep_long).mark_bar().encode(
     x = alt.X("Importance",axis=alt.Axis(labelAngle=-45,labelFontSize=12, title = "Importance"),
              sort = ["Much more","Somewhat more","About the same","Somewhat less","Much less"]),
    y= alt.Y("n_voters", title = "Number of Voters"),
    color = alt.Color("n_voters",scale = alt.Scale(scheme="reds")))

b_text = b.mark_text(dy=-10,align="center").encode(
    text ="n_voters")
bars_and_text = b+b_text

bars_and_text.transform_filter(rep_selection).add_selection(rep_selection).properties(width=600)

alt.LayerChart(...)

alt.VConcatChart(...)